client/tinychain/ml/linalg.py

```diff
@@ -117,6 +117,111 @@ def qr_step(cxt, Q: Tensor, R: Tensor, k: UInt) -> Map:
     return Tensor(QR['Q'])[:cxt.n].transpose(), Tensor(QR['R'])[:cxt.n]
 
 
+class PLUFactorization(Map):
+    """
+    PLU factorization of a given `[N, N]` matrix.
+    """
+    
+    @property
+    def p(self) -> Tensor:
+        """
+        Permutation matrix as an `[N, N]` `Tensor`.
+        """
+        return Tensor(self['p'])
+
+    @property
+    def l(self) -> Tensor:
+        """
+        Lower-triangular matrix as an `[N, N]` `Tensor`.
+        """
+        return Tensor(self['l'])
+
+    @property
+    def u(self) -> Tensor:
+        """
+        Upper-triangular matrix as an `[N, N]` `Tensor`.
+        """
+        return Tensor(self['u'])
+
+@post_op
+def plu(x: Tensor) -> PLUFactorization:
+    """Compute the PLU factorization of the given `matrix`.
+    
+    Args:
+        `x`: a matrix with shape `[N, N]`
+
+    Returns:
+        A `PLUFactorization` instance.
+    """
+
+    class Permutation(Map):
+
+        @property
+        def p(self) -> Tensor:
+            return Tensor(self['p'])
+
+        @property
+        def x(self) -> Tensor:
+            return Tensor(self['x'])
+
+        @property
+        def k(self) -> UInt:
+            return UInt(self['k'])
+
+    def permute_rows(x: Tensor, p: Tensor, start_from: UInt) -> Permutation:
+
+        @post_op
+        def step(p: Tensor, x: Tensor, k: UInt) -> Map:
+            p_k, p_kp1 = p[k].copy(), p[k + 1].copy()
+            x_k, x_kp1 = x[k].copy(), x[k + 1].copy()
+
+            return Map(After(
+                [
+                    p[k].write(p_kp1),
+                    p[k + 1].write(p_k),
+                    x[k].write(x_kp1),
+                    x[k + 1].write(x_k),
+                ],
+                Map(p=p, x=x, k=k + 1)
+            ))
+
+        @post_op
+        def cond(x: Tensor, k: UInt):
+            return (k < UInt(Tuple(x.shape)[0]) - 1).logical_and(x[k, k].abs() < 1e-3)
+
+        result = Map(While(cond, step, Map(
+            p=p.copy(),
+            x=x.copy(),
+            k=start_from,
+        )))
+
+        return Permutation(result)
+
+    @post_op
+    def step(p: Tensor, l: Tensor, u: Tensor, i: UInt) -> Map:
+        pu = Permutation(permute_rows(p=p, x=u, start_from=i))
+        factor = Tensor(pu.x[i + 1:, i] / pu.x[i, i])
+        return Map(After(
+            when=[
+                l[i + 1:, i].write(factor),
+                pu.x[i + 1:].write(pu.x[i + 1:] - factor.expand_dims(axis=0).transpose((1, 0)).copy() * pu.x[i]),
```

Comment:
Also I don't think you need this `copy()`--it looks like this intermediate result is written to `pu.x[i + 1]` exactly once, so there's no advantage to copying it into a new Tensor first